`@nowarn` seems to use "OR" semantics for `&`

## reproduction steps

Scala 2.13.2

Given a warning like:

```
$ scala -e 'def f: Unit = 1'
/tmp/scalacmd15643647702672000142.scala:1: warning: a pure expression does nothing in statement position
def f: Unit = 1
              ^
```

Using `@nowarn("cat=other-pure-statement")` will suppress it:

```
$ scala -e 'import scala.annotation.nowarn; @nowarn("cat=other-pure-statement") def f: Unit = 1'
```

## problem

The problem is that it looks like only one filter need apply:

```
$ scala -e 'import scala.annotation.nowarn; @nowarn("cat=other-pure-statement&msg=bob") def f: Unit = 1'

$ scala -e 'import scala.annotation.nowarn; @nowarn("msg=bob&cat=other-pure-statement") def f: Unit = 1'
```

Compare to `-Wconf` where all filters must apply:

```
$ scala -e 'def f: Unit = 1' '-Wconf:cat=other-pure-statement:silent'

$ scala -e 'def f: Unit = 1' '-Wconf:cat=other-pure-statement&msg=bob:silent'
/tmp/scalacmd13462434343326996307.scala:1: warning: a pure expression does nothing in statement position
def f: Unit = 1
              ^

$ scala -e 'def f: Unit = 1' '-Wconf:msg=bob&cat=other-pure-statement:silent'
/tmp/scalacmd13326564175167716430.scala:1: warning: a pure expression does nothing in statement position
def f: Unit = 1
              ^
```

--- a/src/compiler/scala/tools/nsc/Reporting.scala
+++ b/src/compiler/scala/tools/nsc/Reporting.scala
@@ -612,10 +612,9 @@ object Reporting {
     def used: Boolean = _used
     def markUsed(): Unit = { _used = true }
 
-
     def matches(message: Message): Boolean = {
       val pos = message.pos
-      pos.isDefined && start <= pos.start && pos.end <= end && filters.exists(_.matches(message))
+      pos.isDefined && start <= pos.start && pos.end <= end && filters.forall(_.matches(message))
     }
   }
 }

--- a/test/files/neg/t12026.check
+++ b/test/files/neg/t12026.check
@@ -1,0 +1,6 @@
+t12026.scala:11: warning: discarded non-Unit value
+  def f(): Unit = v()
+                   ^
+error: No warnings can be incurred under -Werror.
+1 warning
+1 error

--- a/test/files/neg/t12026.scala
+++ b/test/files/neg/t12026.scala
@@ -1,0 +1,14 @@
+//scalac: -Wvalue-discard -Werror
+
+import annotation._
+
+//the following option works and warns for f but not g
+//-Wconf:cat=w-flag-value-discard&site=T.g:s
+
+@nowarn("cat=w-flag-value-discard&site=T.g")
+trait T {
+  def v(): Int
+  def f(): Unit = v()
+  def g(): Unit = v()
+}
+// was: no warning

